LunaDll/LuaMain/LuaProxyFFI.cpp

```diff
@@ -568,6 +569,79 @@ typedef struct ExtendedBlockFields_\
         std::wstring wpath = Str2WStr(path);
         return gCachedFileMetadata.exists(wpath);
     }
+
+    FFI_EXPORT(void) LunaLuaSetWindowTitle(const char* newName)
+    {
+        // Add "Software Rendered" if using it, like the normal window name code does
+        std::string fullName = newName;
+        if (gStartupSettings.softwareGL)
+        {
+            fullName += " (Software Renderer)";
+        }
+
+        // Convert the string to the required format
+        // From: https://stackoverflow.com/questions/27220/how-to-convert-stdstring-to-lpcwstr-in-c-unicode
+        // (Not compatible with UTF8? But not sure how easy that'd be to fix or if it's very relevant.)
+        std::wstring stemp = std::wstring(fullName.begin(), fullName.end());
+        LPCWSTR converted = stemp.c_str();
+
+        // Main bit for setting window name
+        SetWindowTextW(gMainWindowHwnd,converted);
+
+        // ... But we also set these three VB-related ones, since otherwise
+        // they seem to cause a weird issue where it can sometimes (and
+        // only sometimes) get reset after refocusing the window.
+        GM_GAMETITLE_1 = fullName;
+        GM_GAMETITLE_2 = fullName;
+        GM_GAMETITLE_3 = fullName;
+    }
+
+    static HICON lastSmallIcon = NULL;
+    static HICON lastBigIcon = NULL;
+
+    FFI_EXPORT(void) LunaLuaSetWindowIcon(LunaImageRef* img, int iconType)
+    {
+        // Convert passed image to a HBITMAP
+        HBITMAP asBitmap = (*img)->asHBITMAP();
+
+        if (asBitmap != nullptr)
+        {
+            // Create an icon out of the image
+            ICONINFO iconInfo;
+
+            iconInfo.fIcon = TRUE;
+            iconInfo.hbmColor = asBitmap;
+            iconInfo.hbmMask = asBitmap;
+            iconInfo.xHotspot = 0;
+            iconInfo.yHotspot = 0;
+
+            HICON asIcon = CreateIconIndirect(&iconInfo);
```

Comment:
The logic for calling `DestroyIcon(...)` isn't quite correct here. Some fairly standard C-style memory management issues. Three problems:
1. You're calling `DestroyIcon` _before_ the `SendMessage` that sets the new icon. This means you're destroying the old icon while it may still be in use, potentially briefly leaving the window referencing an icon handle that's no longer valid. Fixing this should just be a matter of moving the `SendMessage`s above the `DestroyIcon` check.
2. The behavior in the case of `(iconType == 0)` is not correct. You'll end up assigning the same icon handle both `lastSmallIcon` and `lastBigIcon`, and then in a subsequent call you'll end up calling `DestroyIcon` too early while the handle is still in use for the opposite case of big/small.
3. If iconType is for some reason an unexpected value (neither 1, 2, or 3), you leak memory, because you allocated the icon handle but never free it or otherwise keep track of it.

One of the simplest solutions to both 2 and 3 is to move `HICON asIcon = CreateIconIndirect(&iconInfo);` inside your cases for small icon and big icon, ensuring that `lastSmallIcon` and `lastBigIcon` never point to the same icon handle. Alternatively, if you didn't want to do that, you'd have to add a little fancier logic to detect when `(lastSmallIcon == lastBigIcon)` to avoid freeing the icon while it's still referenced by the other variable.

Reply:
This looks to have been fixed properly in [6a16540](https://github.com/WohlSoft/LunaLua/pull/36/commits/6a16540b910e1eadd0266bb2fe919b0b52b7f6dd)